### PR TITLE
UI/Wizard

### PR DESCRIPTION
**Pull request by tonikocjan**
This PR implements a [Wizard](https://docs.microsoft.com/en-us/windows/win32/uxguide/win-wizards)-like UI capability.

In a nutshell, the Wizard provides an animation context, in which different steps of a UI can be animated, and a capability to control the whole flow. The best part is that the view gets automatically resized based on the sizes of steps currently being transitioned.

It is most suitable to be used when you have a multi-step user interface.

https://user-images.githubusercontent.com/15092411/135439993-d59bbab6-f899-401f-930f-6784c09f470d.MP4

https://user-images.githubusercontent.com/15092411/135440020-ded1d6f5-7bc3-4db6-8ffa-fd809632993a.MP4

https://user-images.githubusercontent.com/15092411/135440028-b3f2b184-434b-4f27-881b-5fb8eb09ae5d.MP4

.

### CODE REVIEW

File: Sources/UI/Controllers/Wizard/WizardAnimators.swift [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+  }
+}
+
+public func CompositeAnimator(
+  animators: WizardTransitionAnimator...
+) -> WizardTransitionAnimator {
+  { current, next in
+    animators.forEach { $0(current, next) }
+  }
+}
```
Comment (gentique) [inline Sources/UI/Controllers/Wizard/WizardAnimators.swift:105]: As discussed we should do some form of encapsulation for easier access to these factory functions.

**Review by gentique — APPROVED** [review]

File: Sources/UI/Controllers/Wizard/WizardAnimators.swift (hunk 1 above)
Comment (tonikocjan) [reply Sources/UI/Controllers/Wizard/WizardAnimators.swift]: Please take another look.

File: Sources/UI/Controllers/Wizard/Wizard.swift [hunk 2]
```diff
@@ -0,0 +1,186 @@
...
+  public typealias DataSource = WizardDataSource
+  public typealias LazyStep = DataSource.LazyStep
+  
+  public let contentView: ContentView
+  let dataSource: DataSource
+  
+  public init(
+    steps: [LazyStep],
+    contentView: ContentView
+  ) {
```
Comment (ArlindDushi) [inline Sources/UI/Controllers/Wizard/Wizard.swift:68]: Whose `contentView` is this and why do we need to inject it? Maybe an example/documentation on how this is used would be great 😄

**Review by ArlindDushi — COMMENTED** [review]
This looks promising and has the potential to replace `FittedSheets`. Nice job 👍

File: Sources/UI/Controllers/Wizard/Wizard.swift [hunk 3]
```diff
@@ -0,0 +1,186 @@
+//
+//  Wizard.swift
+//  PovioKit
+//
+//  Created by Toni Kocjan on 29/09/2021.
+//  Copyright © 2021 Povio Labs. All rights reserved.
```
Comment (borut-t) [inline Sources/UI/Controllers/Wizard/Wizard.swift:6]: Update copyright header to year 2022.

File: Sources/UI/Controllers/Wizard/Wizard.swift [hunk 4]
```diff
@@ -0,0 +1,186 @@
...
+//
+
+import Foundation
+import UIKit
+
+public protocol WizardStep: UIViewController {
+}
+
+public class WizardDataSource: NSObject {
```
Comment (borut-t) [inline Sources/UI/Controllers/Wizard/Wizard.swift:15]: Should we split this out into a separate file?

File: Sources/UI/Controllers/Wizard/WizardAnimators.swift [hunk 5]
```diff
@@ -0,0 +1,111 @@
...
+import UIKit
+
+public extension Wizard {
+  enum AnimatorFactory {
+    public typealias Animator = (UIView, UIView) -> Void
+  }
+}
+
+public extension Wizard.AnimatorFactory {
+  static func slideInSlideOut(
```
Comment (borut-t) [inline Sources/UI/Controllers/Wizard/WizardAnimators.swift:18]: A custom implementation could add custom animation as well, right?

File: Sources/UI/Controllers/Wizard/WizardAnimators.swift [hunk 6]
```diff
@@ -0,0 +1,111 @@
+//
+//  WizardAnimators.swift
+//  PovioKit
+//
+//  Created by Toni Kocjan on 29/09/2021.
+//  Copyright © 2021 Povio Labs. All rights reserved.
```
Comment (borut-t) [inline Sources/UI/Controllers/Wizard/WizardAnimators.swift:6]: Update copyright header to year 2022.

File: Sources/UI/Controllers/Wizard/WizardContentView.swift [hunk 7]
```diff
@@ -0,0 +1,107 @@
+//
+//  WizardContentView.swift
+//  PovioKit
+//
+//  Created by Toni Kocjan on 29/09/2021.
+//  Copyright © 2021 Povio Labs. All rights reserved.
```
Comment (borut-t) [inline Sources/UI/Controllers/Wizard/WizardContentView.swift:6]: Update copyright header to year 2022.

File: Sources/UI/Controllers/Wizard/WizardContentView.swift [hunk 8]
```diff
@@ -0,0 +1,107 @@
...
+    }
+    nextTransition(
+      current: current,
+      next: view,
+      transitionDuration: duration,
+      animate: animate,
+      completion: completion)
+  }
+  
+  private func firstTransition(
```
Comment (borut-t) [inline Sources/UI/Controllers/Wizard/WizardContentView.swift:47]: Can we put this into a private extension instead?

**Review by borut-t — CHANGES_REQUESTED** [review]
I think this is a neat, customizable and simple library. Great work 👌 
I would just like to see the following changes:
- we need a documentation/readme about the integration
- update references to "Views" on main readme file

**Comment by borut-t** [comment]
@tonikocjan will you be able to finish up this PR and tackle given comments?

**Pull request closed without merging**